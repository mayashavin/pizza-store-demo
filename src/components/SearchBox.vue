<template>
    <label for="searchbox">Search</label>
    <input v-model="search" placeholder="Search for a pizza" data-testid="search-input" id="searchbox" />
</template>
<script setup>
import { useRouter } from "vue-router";
import { useSearch } from "../composables/useSearch";
import { watch } from "vue";

const props = defineProps({
  searchTerm: {
    type: String,
    required: false,
    default: "",
  },
  items: {
    type: Array,
    required: true,
    default: () => []
  }
});

const emits = defineEmits(['update-items']);

// const route: RouteLocationNormalizedLoaded = useRoute();
const router = useRouter();

const { search, searchResults } = useSearch({
  items: props.items,
  defaultSearch: props.searchTerm,
});

watch(search, (value, prevValue) => {
  if (value === prevValue) return;
  router.replace({ query: { search: value } });
  emits('update-items', searchResults.value)
}, {
    immediate: true
});
</script>