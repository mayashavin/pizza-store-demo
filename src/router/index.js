import {
  createRouter,
  createWebHistory
} from "vue-router";
import PizzasView from "../views/PizzasView.vue";
import ErrorView from "../views/ErrorView.vue";
import PizzaView from "../views/PizzaView.vue";

export const routes = [
  {
    path: "/",
    props: (route) => ({
      searchTerm: route.query?.search || "",
    }),    
    component: PizzasView,
    name: 'home',
  },{
    path: "/:id",
    name: "pizza",
    component: PizzaView,
    props: true,
  },
  {
    path: "/:pathMatch(.*)*",
    name: "error",
    component: ErrorView,
    redirect: "/",
  },
];

export const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes,
});


export default router;