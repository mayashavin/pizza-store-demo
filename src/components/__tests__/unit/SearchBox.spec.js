import { mount } from '@vue/test-utils';
import SearchBox from '@/components/SearchBox.vue';
import { describe, it, expect } from 'vitest'
import { routes } from '@/router';
import { createRouter, createWebHistory } from 'vue-router';

describe("SearchBox component", () => {
    const inputSelector = "[data-testid=search-input]";
    
    const router = createRouter({
        history: createWebHistory(),
        routes: routes,
      })

    it("should get the search term from outside", () => {
        const wrapper = mount(SearchBox, {
            props: {
                searchTerm: "hello",
                items: []
            },
            global: {
                plugins: [router]
            }
        });

        expect(wrapper.find(inputSelector).element.value).toBe("hello");
        expect(wrapper.emitted()).toHaveProperty('update-items')

        const itemsEvent = wrapper.emitted('update-items');

        expect(itemsEvent.length).toBe(1)
    })

    it("should match snapshot", () => {
        const wrapper = mount(SearchBox, {
            props: {
                searchTerm: "",
                items: []
            },
            global: {
                plugins: [router]
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    })
})