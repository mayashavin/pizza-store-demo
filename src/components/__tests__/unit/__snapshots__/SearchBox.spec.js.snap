// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBox component > should match snapshot 1`] = `
"<label for="searchbox">Search</label>
<input placeholder="Search for a pizza" data-testid="search-input" id="searchbox">"
`;
